### PR TITLE
fix(@ngtools/webpack): fix rebuilds for transitive and global type deps

### LINKED ISSUE #15856: "ng serve" doesn't react on changes in d.ts files.
# 🐞 Bug report

### Command (mark with an `x`)

```
- [ ] new
- [ ] build
- [x] serve
- [ ] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] xi18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc
```

### Is this a regression?
I found a bug https://github.com/angular/angular-cli/issues/7638
According to [this](https://github.com/angular/angular-cli/issues/7638#issuecomment-344724819) looks like it worked in angular 5

### Description
"ng serve" doesn't react on changes in d.ts files.

### Precondition:
1. Clone repository: https://github.com/VlasovSergey/angular-ng-serbe-ignores-dts.git
2. run `npm install`
3. run `ng serve`
4. comment `func: () => void;` in file `src\typings.ts`
*Expected:*  'ng serve' runs build and shows an error because `module.ts` uses function func.
*Actual:* 'ng serve' do nothing.

5. re-run `ng serve` (it shows error).
6. make any change in main.ts but don't fix the error (`ng serve` rebuild project and show error correctly)
7. uncomment `func: () => void;` in file `src\typings.ts`

*Expected:* `ng serve` rebuilds the project and the error disappears.
*Actual:* 'ng serve' do nothing. Moreover, if I trigger `ng serve` again (by changing main.ts) the error will be still here.

## 🌍 Your Environment
<pre><code>

Angular CLI: 8.3.9
Node: 10.15.1
OS: win32 x64
Angular: 8.2.11
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.803.9
@angular-devkit/build-angular     0.803.9
@angular-devkit/build-optimizer   0.803.9
@angular-devkit/build-webpack     0.803.9
@angular-devkit/core              8.3.9
@angular-devkit/schematics        8.3.9
@angular/cli                      8.3.9
@ngtools/webpack                  8.3.9
@schematics/angular               8.3.9
@schematics/update                0.803.9
rxjs                              6.4.0
typescript                        3.5.3
webpack                           4.39.2
</code></pre>

### PR DESCRIPTION
**Pull request by filipesilva**
Fix #15856

### CODE REVIEW

File: packages/ngtools/webpack/src/angular_compiler_plugin.ts [hunk 1]
```diff
@@ -609,26 +612,36 @@ export class AngularCompilerPlugin {
...
     //   JS file example `@angular/core/core.ngfactory.ts`.
-    // - .d.ts files might not have a correspondent JS file due to bundling.
     // - __ng_typecheck__.ts will never be requested.
-    const fileExcludeRegExp = /(\.(d|ngfactory|ngstyle|ngsummary)\.ts|ng_typecheck__\.ts)$/;
+    const fileExcludeRegExp = /(\.(ngfactory|ngstyle|ngsummary)\.ts|ng_typecheck__\.ts)$/;
+
+    // Start all the source file names we care about.
+    // Ignore matches to the regexp above, files we've already reported once before, and
+    // node_modules.
+    const sourceFiles = program.getSourceFiles()
```
Comment (alan-agius4) [inline packages/ngtools/webpack/src/angular_compiler_plugin.ts]: Wouldn't it be simpler to do something like?
```ts
this._typeDeps = newTsProgram
      .getSourceFiles()
      .filter(sf => sf.isDeclarationFile)
      .map(({ fileName }) => fileName);
```

 in `_createOrUpdateProgram` which would also result for this fix to be available for VE compilations

Comment (filipesilva) [reply packages/ngtools/webpack/src/angular_compiler_plugin.ts]: Tried adding this at the end of the `_createOrUpdateProgram` method:
```
    console.log(newTsProgram!
      .getSourceFiles()
      .filter(sf => sf.isDeclarationFile && !sf.fileName.includes('node_modules'))
      .map(({ fileName }) => fileName))
```

Then I ran this test. It logged the following:
```
[ 'D:/work/cli/tests/angular_devkit/build_angular/test-project-host-hello-world-app-moqnlwoki9/src/typings.d.ts' ]
[ 'D:/work/cli/tests/angular_devkit/build_angular/test-project-host-hello-world-app-moqnlwoki9/src/typings.d.ts' ]
[ 'D:/work/cli/tests/angular_devkit/build_angular/test-project-host-hello-world-app-moqnlwoki9/src/typings.d.ts' ]
[ 'D:/work/cli/tests/angular_devkit/build_angular/test-project-host-hello-world-app-moqnlwoki9/src/typings.d.ts' ]
[ 'D:/work/cli/tests/angular_devkit/build_angular/test-project-host-hello-world-app-moqnlwoki9/src/typings.d.ts' ]
```

The typings only file was detected, but not the interface files.

**Comment by dgp1130** [comment]
There's a merge conflict attempting to get this into the 9.0.x patch branch. It seems a commit got missed in https://github.com/angular/angular-cli/pull/16148, I'm following up there and once someone confirms, I will include that and merge this PR as well.

**Comment by dgp1130** [comment]
Confirmed that the extra commit in #16148 was an oversight. Merged that into the patch branch 9.0.x and was now able to merge this PR to it as well without any conflicts.

**Pull request merged**